src/main/clojure/clara/rules/compiler.clj

```diff
@@ -386,7 +391,19 @@
 (defn compile-action
   "Compile the right-hand-side action of a rule, returning a function to execute it."
   [binding-keys rhs env]
-  (let [assignments (mapcat #(list (symbol (name %)) (list 'get-in '?__token__ [:bindings %])) binding-keys)
+  (let [;; Avoid creating let bindings in the compile code that aren't actually used in the body.
+        ;; The bindings only exist in the scope of the RHS body, not in any code called by it,
+        ;; so this scanning strategy will detect all possible uses of binding variables in the RHS.
+        ;; Note that some strategies with macros could introduce bindings, but these aren't something
+        ;; we're trying to support.  If necessary a user could macroexpand their RHS code manually before
+        ;; providing it to Clara.
+        rhs-bindings-used (variables-as-keywords rhs)
+
+        assignments (sequence
+                     (comp
+                      (filter rhs-bindings-used)
+                      (mapcat #(list (symbol (name %)) (list 'get-in '?__token__ [:bindings %]))))
```

Comment:
It isn't really part of the change you are making here, but 2 things:

`get-in` shouldn't be left unqualified. There is no good reason to not ensure it is `clojure.core/get-in` instead of perhaps an overridden var with that same name in the compilation ns.

That said, I think using `get-in` here at all is unnecessarily slow. This is a 2 keyword lookup and should be done in a more direct manner.
```clj
(mapcat #(list (symbol (name %)) (-> '?__token__ :bindings %)))
```

Reply:
+1 to both points.  I will make the change.